src/weblets/micro_vm.vue

```diff
@@ -14,71 +14,180 @@
+          ]"
+        >
+          <template #default="{ props }">
+            <v-text-field label="Name" v-model="name" v-bind="props" />
+          </template>
+        </input-validator>
+
+        <SelectVmImage
+          :images="images"
+          v-model:flist="flist"
+          v-model:entry-point="entryPoint"
+        />
+
         <RootFsSize v-model="rootFsSize" />
-        <v-text-field label="CPU (vCores)" v-model.number="cpu" />
-        <v-text-field label="Memory (MB)" v-model.number="memory" />
+
+        <input-validator
+          :value="cpu"
+          :rules="[
+            validators.required('CPU is required.'),
+            validators.isInt('CPU must be a valid integer.'),
+            validators.min('CPU min is 2 cores.', 2),
+            validators.max('CPU max is 32 cores.', 32)
+          ]"
+        >
+          <template #default="{ props }">
+            <v-text-field
+              label="CPU (vCores)"
+              type="number"
+              v-model.number="cpu"
+              v-bind="props"
+            />
+          </template>
+        </input-validator>
+
+        <input-validator
+          :value="memory"
+          :rules="[
+            validators.required('Memory is required.'),
+            validators.isInt('Memory must be a valid integer.'),
+            validators.min('Minimum allowed memory is 256 MB.', 256),
+            validators.max('Maximum allowed memory is 256 GB.', 256 * 1024)
+          ]"
+        >
+          <template #default="{ props }">
+            <v-text-field
+              label="Memory (MB)"
+              type="number"
+              v-model.number="memory"
+              v-bind="props"
+            />
+          </template>
+        </input-validator>
+
         <v-switch color="primary" inset label="Public IPv4" v-model="ipv4" />
         <v-switch color="primary" inset label="Public IPv6" v-model="ipv6" />
         <v-switch color="primary" inset label="Planetary Network" v-model="planetary" />
         <v-switch color="primary" inset label="Add Wireguard Access" v-model="wireguard" />
+
         <SelectFarm
           :filters="{
             cpu,
             memory,
+            publicIp: ipv4,
             ssd: disks.reduce((total, disk) => total + disk.size, rootFsSize)
           }"
           v-model="farm"
+          v-model:country="country"
         />
       </template>
 
       <template #env>
         <ExpandableLayout v-model="envs" @add="envs.push({ key: '', value: '' })">
           <template #default="{ index }">
-            {{ envs[index].key }}
-            <v-text-field label="Key" v-model="envs[index].key" />
+            <input-validator
+                :value="envs[index].key"
+                :rules="[
+                  validators.required('Key name is required.'),
+                  validators.validateKey('Key can\'t start with a number, a non-alphanumeric character or a whitespace'),
```

Comment:
```suggestion
validators.pattern(
 'Key can\'t start with a number, a non-alphanumeric character or a whitespace.', 
 { pattern: /^[a-zA-Z]/ }
),
validators.pattern(
 'Key can\'t start with a number, a non-alphanumeric character or a whitespace.', 
  { pattern: /^[^0-9_\s][a-zA-Z0-9_]+$/ }
  ),
```

You can use validators.pattern to validate a regex it will do the same validation